cmd/keygenerator/main.go

```diff
@@ -206,6 +236,51 @@ func generateKey(keyGen crypto.KeyGenerator, list []key) ([]key, error) {
 	return list, nil
 }
 
+func generateMinedWalletKeys(keyGen crypto.KeyGenerator, list []key, startingHexPattern string, shardID int) ([]key, error) {
+	pattern := nopattern != startingHexPattern
+	withPreferredShard := shardID != noshard && shardID >= 0 && shardID <= 255
+	var patternHexBytes []byte
+	var err error
+	if pattern {
+		patternHexBytes, err = hex.DecodeString(startingHexPattern)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	nbTrials := 0
+	printDeltaTrials := 1000
+	for {
+		if nbTrials%printDeltaTrials == 0 {
+			log.Info("mining address...", "trials", nbTrials)
+		}
+		keys, errKey := generateKey(keyGen, list)
+		if errKey != nil {
+			return nil, errKey
+		}
+		keyBytes := keys[len(keys)-1].pkBytes
+
+		if pattern && !keyHasPattern(keyBytes, patternHexBytes) {
+			nbTrials++
+			continue
+		}
+		if withPreferredShard && !keyInShard(keyBytes, byte(shardID)) {
+			nbTrials++
+			continue
+		}
+		return keys, nil
+	}
+}
+
+func keyHasPattern(key []byte, pattern []byte) bool {
+	return bytes.HasPrefix(key, pattern)
+}
+
+func keyInShard(keyBytes []byte, shardID byte) bool {
+	lastByte := keyBytes[len(keyBytes)-1]
```

Comment:
I think this doesn't work well: for example, if we provide the shardID as 1 only those addresses ending with the value 1 are selected. There are plenty more (like 3 or 5 or ...)
Now, this functionality is not mandatory to be changed as, for example, we want an address that ends with a specific byte in order to be sure that no matter how many shard splits we will do, we end up in the desired shard. This will be most likely used whenever we want to generate an address in the same shard with another one. 
Since shard ID now means a value out of [0-2], we can rename the variable + flag to something like ending byte

Reply:
with 3 or 5 will not work when we have more shards, here we ensure that the address will always be in shard1 (except for when there is only 1 shard - shard 0)